### DISCUSSION TITLE
Linking issues in CentOS 7

### DISCUSSION THREAD

**Discussion by msercheli** [Autodesk/maya-usd#1487] [Q&A] [answered]
I am getting the issue (below) while building maya-usd tag v0.9.0 against:
- CentOS 7
- Maya 2022.0
- devkit (with UFE 2.0.0)
- GCC 6.3.1
- CMake 3.14.6
- USD 20.08 (custom install location)
- Qt 5.15.2 (custom install location)
- Python 2.7.5 (custom install location)

While building MayaUsd plugin, I am running into a CMake warning about library linking.
The build finishes but upon testing loading the plugin in Maya, I get errors of "not finding symbols" (I believe the linking was broken).

Here is the CMake warning (**_these are just 2 examples, but there are many similar warnings_**):

```
CMake Warning at lib/mayaUsd/CMakeLists.txt:18 (add_library):
  Cannot generate a safe linker search path for target mayaUsd because there
  is a cycle in the constraint graph:

    dir 0 is [/usr/autodesk/maya2022/lib]
      dir 1 must precede it due to link library [libtbb.so]
    dir 1 is [/home/dev/usd-20.08/lib]
      dir 0 must precede it due to link library [libtbb.so]

  Some of these libraries may not be found correctly.
```

```
CMake Warning at plugin/adsk/plugin/CMakeLists.txt:18 (add_library):
  Cannot generate a safe runtime search path for target mayaUsdPlugin because
  there is a cycle in the constraint graph:

    dir 0 is [/home/dev/devkit-2022.0/devkit/ufe/lib]
    dir 1 is [/home/dev/build_dir/lib/usd/pxrUsdPreviewSurface]
    dir 2 is [/home/dev/build_dir/lib/usd/ui]
    dir 3 is [/home/dev/build_dir/lib/mayaUsd]
    dir 4 is [/home/dev/usd-20.08/lib]
      dir 6 must precede it due to runtime library [libtbb.so]
    dir 5 is [/home/dev/usd-20.08/lib64]
    dir 6 is [/usr/autodesk/maya2022/lib]
      dir 0 must precede it due to runtime library [libufe_2.so]
      dir 4 must precede it due to runtime library [libosdCPU.so.3.4.3]
      dir 5 must precede it due to runtime library [libPtex.so]
      dir 8 must precede it due to runtime library [libpython2.7.so.1.0]
      dir 9 must precede it due to runtime library [libQt5Widgets.so.5]
    dir 7 is [/home/dev/build_dir/lib/usd/utils]
    dir 8 is [/home/dev/python-2.7.5//lib]
    dir 9 is [/home/dev/qtbase-5.15.2/python-2.7/lib]

  Some of these libraries may not be found correctly.
```

It seems like, since I am passing to the plugin build system the `pxrusd-location`, `maya-location` and `devkit-location`, it is getting confused locating the proper library to link.

There are no linking errors during the build, only while loading the plugin in Maya.

Appending the `usd-location` lib directory to the `LD_LIBRARY_PATH` doesn't help. In fact, Maya spits Python import errors for the USD modules.

I could be wrong but I believe Maya 2022 comes with USD since I can `import pxr.Usd` from the ScriptEditor.
Although I could not find any USD lib or include header files within the Maya install location, I was wondering if I should be building the plugin against the USD that is shipped with Maya 2022?

**Comment by HamedSabri-adsk** [comment, marked as answer]
@msercheli 

> I am running into a CMake warning about library linking.

You can ignore the warning "cycle in the constraint graph". You are getting the warning since we are linking against same libraries from different directories at the same time.  In your example, there are two libtbb.so one in "/usr/autodesk/maya2022/lib" and the other in "/home/dev/usd-20.08/lib"

> Appending the usd-location lib directory to the LD_LIBRARY_PATH doesn't help. 

You shouldn't need to manually set LD_LIBRARY_PATH at all.

It seems like you have a path to another USD or another Python in your path. We have seen this issue in the past. Try running following commands to be sure only the USD that you have passed to pxrusd-location is in your path. 

getenv "PATH";
getenv "PYTHONPATH";

> I could be wrong but I believe Maya 2022 comes with USD since I can import pxr.Usd from the ScriptEditor.

When you build maya-usd, you are building 3 plugins:  Maya Usd, Animal Logic Usd, and Pixar usd.

The reason you can import pxr.Usd is because, the PATH and PYTHONPATH for Pixar USD ( the one you passed to --pxrusd-location ) is set automatically.

https://github.com/Autodesk/maya-usd/blob/06ffd434fc02e01c42e700c2f5c6e8746c0f927f/cmake/usd.cmake#L16
 
> I was wondering if I should be building the plugin against the USD that is shipped with Maya 2022

AFAIK, Pixar USD include/libs are not shipped with any versions of Maya.

**Reply by msercheli** [reply]
Hi Hamed,

Thanks for the answer.
I believe Maya 2022 is shipping with the USD libraries (at least for Linux). I can find them in:
`/usr/autodesk/mayausd/maya2022/0.8.0_202102180129-2f83c8f/mayausd`, side-by-side with the Maya installation itself.

There, I can find the Maya-Usd plugin for both Python 2 and 3 as well as USD itself for Python 2 and 3.
Although the includes for USD are there, they are zipped (devkit.zip).

I will double check my PATH and PYTHONPATH to be sure I only have the USD provided to the plugin.
In relation to Python, I have another Python in my system. Are you saying that, while building the plugin, I should only use Python provided by Maya (mayapy)?

Once again, thanks for the help.

**Comment by HamedSabri-adsk** [comment]
Hi Marcelo, 

I was referring to Pixar USD ( the core library itself ) when you mentioned `Maya 2022 comes with USD`. I thought when you said USD, you were referring to Pixar USD ( core library ).

You don't necessarily need to use `mayapy` to build the plugins. You can use your system python.  

`"Your system python (either 2.x or 3.x)" build.py --maya-location "" --pxrusd-location "" .....`

However, it is important that Python version that MayaUSD is built against matches the Python version that Maya use. By default, maya-usd projects is built with Python 2.

https://github.com/Autodesk/maya-usd/blob/06ffd434fc02e01c42e700c2f5c6e8746c0f927f/CMakeLists.txt#L35

I believe Maya on Windows and Linux now starts in Python 3 mode by default:

https://help.autodesk.com/view/MAYAUL/2022/ENU/?guid=GUID-DF43840B-4DB1-43F8-BFD1-97D8D031B91D

If you want to build Maya-USD with Python3 then you need to turn on the `BUILD_WITH_PYTHON_3` flag

e.g
`build.py --maya-location "" --pxrusd-location "" --build-args=-DBUILD_WITH_PYTHON_3=ON
`

**Reply by msercheli** [reply]
Hi Hamed,

Sorry for the confusion. I was actually referring to the core USD libraries/includes from Pixar.
Actually, although the name of the directory may be misleading, the Pixar core USD libs and includes (zipped) are there.
Here is the contents from my Maya 2022 installation (raw) directory. The `mayausd` subdirectory you see here is the one shipping with 2022 out-of-the-box:

```
/usr/autodesk $ ls -l
total 4
drwxrwxr-x.  3 root root   22 May 13 18:15 arnold
drwxr-xr-x.  3 root root   22 May 13 18:16 bifrost
lrwxrwxrwx.  1 root root    8 May 13 18:15 maya -> maya2022
drwxr-xr-x. 31 root root 4096 May 13 18:15 maya2022
drwxr-xr-x.  3 root root   22 May 13 18:15 mayausd
drwxrwxr-x.  3 root root   18 May 13 18:15 modules
drwxr-xr-x.  3 root root   19 May 13 18:15 pymel2022
```

Going deeper into the `mayausd` directory:
```
/usr/autodesk/mayausd/maya2022/0.8.0_202102180129-2f83c8f/mayausd $ ls -l
total 0
drwxr-xr-x. 5 root root  46 May 13 18:15 MayaUSD2
drwxr-xr-x. 5 root root  46 May 13 18:15 MayaUSD3
drwxr-xr-x. 9 root root 165 Jun 14 11:06 USD2
drwxr-xr-x. 9 root root 165 Jun 14 10:55 USD3
```

Going inside the USD3 for instance (but the same subdirectories can be found in USD2):
```
/usr/autodesk/mayausd/maya2022/0.8.0_202102180129-2f83c8f/mayausd/USD3 $ ls -l
total 36912
drwxr-xr-x.  2 root root     4096 May 13 18:16 bin
drwxr-xr-x.  2 root root      142 Jun 14 10:54 cmake
-rw-r--r--.  1 root root 37763215 Feb 18 05:04 devkit.zip
drwxr-xr-x. 10 root root     4096 Jun 14 10:54 include
drwxr-xr-x.  5 root root     8192 May 13 18:16 lib
drwxr-xr-x.  2 root root       87 May 13 18:16 lib64
drwxr-xr-x.  3 root root       17 May 13 18:16 plugin
-rw-r--r--.  1 root root     1229 Feb 18 05:04 pxrConfig.cmake
-rw-r--r--.  1 root root     1229 Feb 18 05:04 pxrConfig.cmake.orig
drwxr-xr-x.  5 root root      103 Jun 14 10:55 share
```

The subdirectories `include`, `share`, `cmake` were unzipped from the `devkit.zip` file.

I am going to try something that I didn't before. I am going to try and build the Maya-Usd plugin against the USD location (USD2 and USD3) above and using `mayapy2` (or `mayapy` for building using Python3). I just want to see if it works.

**Comment by msercheli** [comment]
Concerning the linking issues, since Hamed mentioned that the linking warnings can be ignored, I am marking this discussion as answered.
I am able to build the plugin, but it still gives me error while attempting to load it into Maya. I will create a new discussion for the error to keep them separated.
